Externally commit script utxo

ffakenz opened the pull request:
This PR adds the option for our users to commit a script utxo to the Head externally, enabling script utxos to be made available in a Hydra head.


:palm_tree: Extend the HTTP endpoint data types to accept utxo with an optional script `witness` (like datum, redeemer and script)

:palm_tree: Generalize the Chain handler `draftTx` to build, balance and sign internally the drafted tx using utxos with script witnesses.

:palm_tree: Reject commit transaction requests including UTxO addressed to the internal wallet.

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

ffakenz committed:
Create another external commit scenario for scripts

This is using the drafted scenario created by @ffakenz and @v0d1ch

ch1bo committed:
WIP: draft a spending tx for script

To test things outside hydra-node

ffakenz committed:
Follow up on temporary sanity check: Spend the script on L1

- Replace random script with dummy always succeed.
- Fix script datum to be ScriptDatumForTxIn.
- Add missing protocol params and collateral to script tx.
- Fix script output to not use TxOutDatumNone as it is invalid.

Note: This still fails because of InsufficientCollateral

ffakenz committed:
Fix script tx to be spent on L1

- Add missing collateral
- Add missing witness during signing

Note:
By introducing the collateral utxo, now the key has more than one.
This causes the resulting utxo from queryUTxOFor contained two entries: one for
the normal utxo and another one for the collateral utxo.
This was making createScriptOutput to build an invalid script tx in case the
collateral utxo was selected, when finding a utxo that fits the total amount
to deposit.

Finally, removed debug lines

ffakenz committed:
Add script info as part of draft commit tx request

The script info is necessary to build a script tx.

the new draftScriptTx chain handler is a dummy implementation and has the same
behavior as draftTx.

ffakenz committed:
Add protocol parameters, systemStart and epochInfo as arguments to mkChain factory method.

They are necessary to draft script txs.

ffakenz committed:
Minor formatting change

ffakenz committed:
Split commitTx into tx body builder and tx builder

Added commitScript tx builder draft with FIXME note.

ffakenz committed:
Finish commit external script e2e spec

Removed systemStart and epochInfo as arguments to mkChain factory method as they got redundant.

ffakenz committed:
Minor formatting changes

ffakenz committed:
Fix golden files for rest server

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -192,6 +232,105 @@ singlePartyCommitsFromExternal tracer workDir node@RunningNode{networkId} hydraS
+      let reedemer = toScriptData () -- ScriptDataBytes mempty
+          datum = toScriptData () -- ScriptDataBytes mempty
+          collateralTxIns = fst <$> UTxO.pairs colateralUTxO
+          scriptInfo = ScriptInfo reedemer datum script collateralTxIns
```
As we are implementing option A, I think the collateral can come from the hydra-node. Similarly as the utxo to pay fees.

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -192,6 +232,105 @@ singlePartyCommitsFromExternal tracer workDir node@RunningNode{networkId} hydraS
+      let DraftCommitTxResponse commitTx = responseBody response
+
+      -- sign and submit the tx with our external user key
+      let signedCommitTx = signWith someSk commitTx
```
Signing should not be necessary (if we use a node-owned collateral)

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -192,6 +232,105 @@ singlePartyCommitsFromExternal tracer workDir node@RunningNode{networkId} hydraS
+        pparams
+        scriptAddress
+        mempty
+        (mkTxOutDatumHash ())
```
This must match the datum later provided, pass it into this function.

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -192,6 +232,105 @@ singlePartyCommitsFromExternal tracer workDir node@RunningNode{networkId} hydraS
+      (someVk, someSk) <- generate genKeyPair
+      pparams <- queryProtocolParameters networkId nodeSocket QueryTip
+      normalUTxO <- seedFromFaucet node someVk 10_000_000 Normal (contramap FromFaucet tracer)
+      scriptUtxo <- createScriptOutput pparams scriptAddress someSk normalUTxO
```
Could skip the `seedFromFaucet` by creating the output directly using the faucet key and giving change to the faucet.

ch1bo commented on hydra-node/src/Hydra/Chain.hs:
```diff
@@ -186,6 +191,14 @@ data Chain tx m = Chain
+      ScriptRedeemer ->
+      PlutusScript ->
+      [TxIn] ->
+      m (Either (PostTxError tx) tx)
```
Don't create another function, but instead generalize the `draftTx`.

ch1bo commented on hydra-node/src/Hydra/API/RestServer.hs:
```diff
@@ -47,19 +56,33 @@ instance IsTx tx => Arbitrary (DraftCommitTxResponse tx) where
+
+data DraftCommitTxRequest tx = DraftCommitTxRequest
+  { utxo :: UTxOType tx
+  , scriptInfo :: Maybe ScriptInfo
```
This is not enough. What if we want to spend from two script outputs?

We should aim for a type similar to `cardano-api`'s `TxIns`, for example
```
toCommit :: [(TxIn, TxOut, Maybe ScriptWitness)]
```

And .. it should not be cardano-specific, which makes this a bit annoying. For example using type families like `TxOutType tx` etc.

ch1bo reviewed (commented):
Leaving some comments to discuss together

ffakenz committed:
Make collateral to come from the node

- Remove collateralTxIns argument from draftScriptTx handler interface.
- Now the collateral comes from the node initial state (initialThreadUTxO).
- This removes the collateralTxIns attribute from script info request payload.
- This also simplifies the client side as there is no need to sign a drafted tx.

v0d1ch committed:
Remove extra param to draftScriptTx

v0d1ch committed:
Commit from two script utxos

v0d1ch committed:
WIP: Start to model proper type for committing multiple scripts

ffakenz committed:
Fix collateralTxIns to use the internal wallet utxo instead of the head inital

ffakenz committed:
Alter api to be able to draft a commit using multiple scripts

Note: this is failing because of ErrTranslationError.
This error is thrown while calling evaluateTransactionExecutionUnits as part of
coverFee during finalizeTx on draftScriptTx.

v0d1ch committed:
WIP: Working scripts commit with mempty user utxo

ffakenz committed:
WIP: Working scripts commit with mempty user utxo

Enrich draft commit tx with:
- collateral TxIns
- protocol params
- TxIns with Witnesses for utxo scripts

ffakenz committed:
Generalize draftTx to work both for regular and script utxo

ffakenz committed:
Check we can draft a commit tx using many regular and many script utxos

In order to submit a tx drafted using a regular utxo it needs to be signed by
the utxo key before.

ffakenz committed:
Remove collateralTxIns and pparams from draftTx

This is already done during finalizeTx.
This also removes the need to pass the pparams to mkChain.

Also removed the error handling on draftTx leaving it to the Server.

ffakenz committed:
Unify types in draft commit tx request to not be cardano specific

v0d1ch committed:
WIP: Remove tags from http server types

Problems arize when we try to expres the draft request type
in the api.yaml file. How to express a list of tuples?

WIP: Remove tags from http server types

Problems arize when we try to expres the draft request type
in the api.yaml file. How to express a list of tuples?

WIP: continue working on api yaml

v0d1ch committed:
Reject commit transaction requests including UTxO addressed to the internal wallet

ffakenz committed:
Add tag to request as it is needed to validate the api schema.

ffakenz committed:
Add plain old json schema validation for draft commit request

pgrange commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -192,6 +237,158 @@ singlePartyCommitsFromExternal tracer workDir node@RunningNode{networkId} hydraS
+-- | Single hydra-node where the commit is done from an external UTxO owned by a
+-- script which requires providing script, datum and redeemer instead of
+-- signing the transaction.
+singlePartyCommitsFromExternalScript ::
```
I would make this test a _BehaviorSpec_ test so that there is less plumbing involved and keep only a very basic test here.

v0d1ch replied:
I am not sure we can test the same thing at this level. AFAIS `BehaviourSpec` is using `SimpleTx` and mocks the chain while in the end-to-end we have the api server available thus testing is easy there. We think there is no _behaviour_ to test here, we need to have the chain, api and the ledger available and I am not seeing the easy way to do it here.

pgrange commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -192,6 +237,158 @@ singlePartyCommitsFromExternal tracer workDir node@RunningNode{networkId} hydraS
+        (mkTxOutDatumHash datum)
+        ReferenceScriptNone
+
+singlePartyCantCommitExternallyWalletUtxo ::
```
I would make this test a _BehaviorSpec_ test so that there is less plumbing involved and keep only a very basic test here.

v0d1ch replied:
See https://github.com/input-output-hk/hydra/pull/903#discussion_r1229400026

pgrange commented on hydra-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -132,11 +133,21 @@ spec = around showLogsOnFailure $ do
             publishHydraScriptsAs node Faucet
               >>= timedTx tmpDir tracer node
-      it "commits from external wallet" $ \tracer -> do
+      fit "commits from external utxo" $ \tracer -> do
```
```suggestion
      it "commits from external utxo" $ \tracer -> do
```

pgrange commented on hydra-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -132,11 +133,21 @@ spec = around showLogsOnFailure $ do
           withCardanoNodeDevnet (contramap FromCardanoNode tracer) tmpDir $ \node ->
             publishHydraScriptsAs node Faucet
               >>= singlePartyCommitsFromExternal tracer tmpDir node
+      fit "commits from external script utxo" $ \tracer -> do
```
```suggestion
      it "commits from external script utxo" $ \tracer -> do
```

pgrange commented on hydra-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -132,11 +133,21 @@ spec = around showLogsOnFailure $ do
             publishHydraScriptsAs node Faucet
               >>= timedTx tmpDir tracer node
-      it "commits from external wallet" $ \tracer -> do
+      fit "commits from external utxo" $ \tracer -> do
```
I would make this test a BehaviorSpec test so that there is less plumbing involved and keep only a very basic test here.

ffakenz replied:
same thread as [above](https://github.com/input-output-hk/hydra/pull/903#discussion_r1226646988)
lets continue it there.

pgrange commented on hydra-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -132,11 +133,21 @@ spec = around showLogsOnFailure $ do
+          withCardanoNodeDevnet (contramap FromCardanoNode tracer) tmpDir $ \node ->
+            publishHydraScriptsAs node Faucet
+              >>= singlePartyCommitsFromExternalScript tracer tmpDir node
+      it "can't commit externally internal wallet utxo" $ \tracer -> do
```
I would make this test a BehaviorSpec test so that there is less plumbing involved and keep only a very basic test here.

v0d1ch replied:
See https://github.com/input-output-hk/hydra/pull/903#discussion_r1229400026

pgrange commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -255,11 +255,11 @@ components:
             type: string
             enum: ["GetUTxO"]
-
+    
```
```suggestion
 
```

pgrange commented on hydra-node/src/Hydra/API/RestServer.hs:
```diff
@@ -20,45 +24,70 @@ deriving stock instance IsTx tx => Show (DraftCommitTxResponse tx)
     object
-      [ "tag" .= String "DraftCommitTxRequest"
+      [ -- XXX: tag should be not needed but is required
+        -- to validate its api aschema
```
```suggestion
        -- to validate its api schema
```

ffakenz replied:
outdated, as tag was removed from request.

pgrange commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -1245,6 +1243,18 @@ components:
             tag:
               type: string
               enum: ["FailedToDraftTxNotInitializing"]
+        - title: FailedToDraftTxWalletUtxoDetected
```
```suggestion
        - title: SpendingNodeUtxoForbidden
```

The user is already in a context where they try to draft a tx so I would remove _FailedToDraft_ from the name.

Since we detected Utxo from the node's wallet, I would make it clear that it is forbidden to try to spend this UTXO.

ch1bo replied:
No strong opinion about the name, but much more important: this schema is unused!

Must: mention this (and other errors) as a possible response on the `/commit` endpoint.

pgrange commented on hydra-node/src/Hydra/Chain/Direct/Handlers.hs:
```diff
@@ -145,11 +162,52 @@ mkChain tracer queryTimeHandle wallet ctx LocalChainState{getLatest} submitTx =
         case Hydra.Chain.Direct.State.chainState chainState of
-          Initial st ->
-            sequenceA $ finalizeTx wallet ctx chainState utxo <$> commit ctx st utxo
+          Initial st -> do
```
I would try to reduce this function length. Maybe there's a way to extract some functions and keep it clear?

pgrange commented on hydra-node/src/Hydra/Chain/Direct/State.hs:
```diff
@@ -313,57 +316,75 @@ commit ctx st utxo = do
-    , initialHeadTokenScript
-    , headId
-    } = st
+-- | Construct a commit a script transaction based on the 'InitialState'.
```
```suggestion
-- | Construct a commit script transaction based on the 'InitialState'.
```

pgrange commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -178,12 +178,28 @@ commitTx ::
-      & addVkInputs committedTxIns
-      & addExtraRequiredSigners [vkh]
-      & addOutputs [commitOutput]
+    rawCommitTxBody networkId scriptRegistry headId party utxo (initialInput, out, vkh)
```
I would keep the function code inline as it was before as I don't see more clarity in the code by this particular function extraction.

ffakenz replied:
This function has been extracted in order to be re-used when drafting a commit using a script utxo, not to add improve clarity in the code.

ch1bo replied:
@pgrange has the right gut feeling. This is a major smell :hankey:. `commitTx` is a pure function creating a commit transaction, it's only call site `commit` in `State` module is basically a thin wrapper, still pure and producing a `Tx`. So we should use it. See my other comments for details.

ch1bo committed:
WIP: Draft a prop_validateJSONSchema function

This could be used validate parts of the api spec (the included JSON
schemas) against serialization of Haskell types.

ffakenz committed:
Remove tag from request + Fix api schema validation for /commit

- Do not use async-api validator for /commit request and response, instead use a
plain old json schema validation (prop_validateJSONSchema).
- Remove /commit request and response from api.yaml#/components/messages because
/commit#subscribe and /commit#publish message cannot be a reference object but a
message object instead, due to message objects contains a payload that is a
valid JSON schema.
- The use of  in JSON schemas is needed to make sure the validator fails for the
wrong reason.
- Note the enrichment of jsonSchema during prop_validateJSONSchema by adding an
id field and wrapping the schema around an array type.

WIP: there are improvements left todo marked as XXX

ch1bo commented on hydra-node/src/Hydra/Chain.hs:
```diff
@@ -185,7 +191,10 @@ data Chain tx m = Chain
-  , draftTx :: (IsChainState tx, MonadThrow m) => UTxOType tx -> m (Either (PostTxError tx) tx)
+  , draftTx ::
+      (IsChainState tx, MonadThrow m, MonadIO m) =>
+      [(UTxOType tx, Maybe (ScriptDatum WitCtxTxIn, ScriptRedeemer, PlutusScript))] ->
```
This is a suboptimal type. It allows for more variability as we need it. i.e. multiple utxos can be either multiple list entries, or one list entry with multiple utxos in the first position.

Must: use `[(TxIn, TxOut CtxUTxO, Maybe (ScriptDatum WitCtxTxIn, ScriptRedeemer, PlutusScript))]`

ch1bo commented on hydra-node/src/Hydra/API/RestServer.hs:
```diff
@@ -20,62 +24,80 @@ deriving stock instance IsTx tx => Show (DraftCommitTxResponse tx)
+
 newtype DraftCommitTxRequest tx = DraftCommitTxRequest
-  { utxos :: UTxOType tx
+  { utxos :: [DraftUTxO tx]
```
This is a suboptimal type. It allows for more variability as we need it. i.e. multiple utxos can be either multiple list entries, or one list entry with multiple utxos as `draftUTxO`.

**Must:** change this type t something like

```
newtype DraftCommitTxRequest tx = DraftCommitTxRequest
  { utxos :: [(TxIn, TxOut CtxUTxo, Maybe ScriptInfo)]
  }
```

or update `DraftUTxO` consequently.

ch1bo commented on hydra-node/src/Hydra/API/RestServer.hs:
```diff
@@ -1,4 +1,5 @@
 {-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
```
Should: not use this pragma unless really needed. Can't see an orphan instance in this module.

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -1601,6 +1602,54 @@ components:
+    PlutusV2Script:
+      type: object
+      description: |
+        Plutus V2 Script data
```
This is in fact a "TextEnvelope" of a PlutusV2Script. See also `TextEnvelopeTransaction`.

Should: not make `description` as required and disallow `additionalProperties`

ch1bo commented on hydra-node/src/Hydra/API/RestServer.hs:
```diff
@@ -20,62 +24,80 @@ deriving stock instance IsTx tx => Show (DraftCommitTxResponse tx)
-      [ "tag" .= String "DraftCommitTxRequest"
-      , "utxos" .= toJSON utxo
+      [ "utxos" .= toJSON utxo
       ]
```
Should: avoid writing instances by hand

ch1bo commented on hydra-node/src/Hydra/Chain.hs:
```diff
@@ -185,7 +191,10 @@ data Chain tx m = Chain
   --
   -- Does at least throw 'PostTxError'.
-  , draftTx :: (IsChainState tx, MonadThrow m) => UTxOType tx -> m (Either (PostTxError tx) tx)
+  , draftTx ::
```
Should: name this `draftCommitTx` as it is only about commit transactions right now.

ch1bo commented on hydra-node/src/Hydra/API/Server.hs:
```diff
@@ -350,17 +353,29 @@ handleDraftCommitUtxo directChain tracer body reqMethod reqPaths respond = do
+                Nothing ->
+                  (utxo, Nothing)
+          )
+            <$> utxos
```
If we use different types in the API (makes sense because of json instances), do create a conversion function next to `DraftCommitTxRequest` along the lines of `convertDraftUTxO :: [DraftUTxO] -> [(TxIn, TxOut CtxUTxo, ...)]` (depending on the type used by `draftTx`)

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Handlers.hs:
```diff
@@ -15,17 +16,31 @@ import qualified Cardano.Api.UTxO as UTxO
   getTxBody,
   getTxId,
+  mkScriptWitness,
+  pattern ScriptWitness, fromLedgerTxIn,
```
File not formatted by fourmolu?

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Handlers.hs:
```diff
@@ -145,11 +162,52 @@ mkChain tracer queryTimeHandle wallet ctx LocalChainState{getLatest} submitTx =
+              , BuildTxWith $
+                  ScriptWitness ScriptWitnessForSpending $
+                    mkScriptWitness script datum redeemer
+              )
```
This indicates that we should use `ScriptWitness` already on the type signature of `draftTx`

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Handlers.hs:
```diff
@@ -145,11 +162,52 @@ mkChain tracer queryTimeHandle wallet ctx LocalChainState{getLatest} submitTx =
+              (u, d, r, s) <- scriptUtxoInput
+              let txIns = Set.toList $ UTxO.inputSet u
+                  scriptUtxoInfos = (,d,r,s) <$> txIns
+              pure scriptUtxoInfos
```
Is this function unused?

ffakenz replied:
yes it is.

ch1bo commented on hydra-node/test/Hydra/API/RestServerSpec.hs:
```diff
@@ -17,3 +21,27 @@ spec = parallel $ do
+            ]
+  aroundAll withJsonSpecifications $
+    -- XXX: It's unintuitive that dir/api.yaml is in fact a JSON file holding
+    -- the JSON schema!
```
Should: not repeat `aroundAll` (and the comment)

ch1bo commented on hydra-node/test/Hydra/API/RestServerSpec.hs:
```diff
@@ -17,3 +21,27 @@ spec = parallel $ do
+        withMaxSuccess 1 $ do
+          conjoin
+            [ prop_validateJSONSchema @(DraftCommitTxResponse Tx) (dir </> "api.yaml") (key "channels" . key "/commit" . key "subscribe" . key "message" . key "payload")
+            , prop_specIsComplete @(ReasonablySized (DraftCommitTxResponse Tx)) (dir </> "api.yaml") apiSpecificationSelector
```
Should: not need `ReasonablySized` for the responses

ch1bo commented on hydra-node/test/Hydra/API/RestServerSpec.hs:
```diff
@@ -17,3 +21,27 @@ spec = parallel $ do
+        withMaxSuccess 1 $ do
+          conjoin
+            [ prop_validateJSONSchema @(DraftCommitTxRequest Tx) (dir </> "api.yaml") (key "channels" . key "/commit" . key "publish" . key "message" . key "payload")
+            , prop_specIsComplete @(ReasonablySized (DraftCommitTxRequest Tx)) (dir </> "api.yaml") apiSpecificationSelector
```
Should: not need `ReasonablySized` here

ffakenz replied:
below [comment](https://github.com/input-output-hk/hydra/pull/903#discussion_r1227715180) requests to remove this spec as it was pointless.

ch1bo commented on hydra-node/test/Hydra/API/RestServerSpec.hs:
```diff
@@ -17,3 +21,27 @@ spec = parallel $ do
+        withMaxSuccess 1 $ do
+          conjoin
+            [ prop_validateJSONSchema @(DraftCommitTxRequest Tx) (dir </> "api.yaml") (key "channels" . key "/commit" . key "publish" . key "message" . key "payload")
+            , prop_specIsComplete @(ReasonablySized (DraftCommitTxRequest Tx)) (dir </> "api.yaml") apiSpecificationSelector
```
This spec is pointless as the request has only a single instance and there is no need to check for completeness.

ch1bo commented on hydra-node/test/Hydra/API/RestServerSpec.hs:
```diff
@@ -17,3 +21,27 @@ spec = parallel $ do
+        withMaxSuccess 1 $ do
+          conjoin
+            [ prop_validateJSONSchema @(DraftCommitTxResponse Tx) (dir </> "api.yaml") (key "channels" . key "/commit" . key "subscribe" . key "message" . key "payload")
+            , prop_specIsComplete @(ReasonablySized (DraftCommitTxResponse Tx)) (dir </> "api.yaml") apiSpecificationSelector
```
In principle checking completeness of possible messages on the response side would make sense, this property is nonsense given the current `api.yaml`.

See also my comment that we should enumerate the error responses in the api specification.

v0d1ch replied:
I tried enumerating all possible outcome messages comming out of the draft response without success. The error messages we return to the users are a subset of `PostTxError` (+ the 500 ones) and whatever I try I can't reference only some messages as part of this response. Let's pair on it @ch1bo

ch1bo commented on hydra-node/test/Hydra/JSONSchema.hs:
```diff
@@ -23,17 +23,65 @@ import System.Exit (ExitCode (..))
+-- which selects the JSON schema for "Address" types in a bigger specification,
+-- say an asyncapi description.
+--
+-- TODO: Maybe do the same trick as prop_validateToJSON to avoid slow runs.
```
We do this. Should remove the TODO

ch1bo reviewed (changes requested):
Have not reviewed 100%, but as there are already some major things to address and their changes rippling through the way other code will be written I leave it here.

There is also lots of things to clean up and it's a bit annoying needing to point out leftovers as a reviewer.

ffakenz committed:
Fix logs and rest api schema validation.

- Now if the assertion fails, we will keep a dir under /tmp containing the
failing context.
- Note we changed the structure of the property tests because there is no
 anymore as each property will call
 and be responsible for their own /tmp dir.

v0d1ch committed:
PR review: Alter the RestServer DraftUTxO type

ffakenz committed:
Apply review request

- Rename handler draftTx to draftCommitTx

- Remove unnecessary abstract type tx to request and draft utxo types.

- Extract convertDraftUTxO logic from handleDraftCommitUtxo as a helper function

- Also remove spec to check for completeness on request as it has only a single
instance

- Improve imports

- Fix minor typos

- Remove focus on e2e specs

ffakenz committed:
Simplify draftCommitTx handler signature to take ScriptWitness instead of the triplet (datum, redeemer, script)

ffakenz committed:
Refactor draftCommitTx to be simpler as per request for review

v0d1ch committed:
Alter PlutusV2Script api entry

v0d1ch committed:
Do not use ReasonablySized for DraftCommitTxResponse test

v0d1ch committed:
Rename FailedToDraftTxWalletUtxoDetected to SpendingNodeUtxoForbidden

- Also update the api.yaml to include possible error responses

v0d1ch committed:
Simplify the way we create script outputs

v0d1ch committed:
Remove api.yaml mention from the file itself

- Alter the RestServerSpec path in tests

v0d1ch committed:
Do not validate http endpoints and re-generate DraftCommitTxResponse
json

v0d1ch committed:
Improve on api docs display and add missing docusaurus logger

v0d1ch committed:
Don't pass faucet pub key, query for it inside of a function.

v0d1ch committed:
Reduce temp folder name length

v0d1ch committed:
Update schema path

v0d1ch committed:
Rebase fix

ch1bo commented on hydra-node/golden/ReasonablySized (DraftCommitTxRequest (Tx BabbageEra)).json:
```diff
@@ -1,386 +1,357 @@
+                            "lovelace": 13
+                        }
                     }
                 },
```
Instead of this quite different serialization, I think we should aim for something similar to `UTxO`, but extended by an optional `"witness"` field. For example:

```
{
  "utxoToCommit": {
    "0103050804080403050205060701050704000503030006020506030804080506#71": {
      "address": "addr_test1wrs2w9p3nqfv8amnhgzwchtt8l7dt2kc2qrgqkcy0vyz2sg3p9azg",
      "datum": null,
      "inlineDatum": {
        "bytes": ""
      },
      "inlineDatumhash": "39df024ac52722fe8ae4c1a8740e4c5624a38c3820e504a059aae8728421f8bd",
      "referenceScript": null,
      "value": {
        "1920e782f048e9ef52acd89c4341a89c3908c6e046ad87c474fffb48": {
          "0b0a00100500100c": 108
        },
        "3542acb3a64d80c29302260d62c3b87a742ad14abf855ebc6733081e": {
          "0010061010": 48
        },
        "65fc709a5e019b8aba76f6977c1c8770e4b36fa76f434efc588747b7": {
          "0606040f10050c0b0b": 64
        },
        "lovelace": 13
      },
      "witness": {
        "plutusV2Script": {
          "cborHex": "40",
          "description": "",
          "type": "PlutusScriptV2"
        },
        "datum": "06",
        "redeemer": "23"
      }
    }
  }
}

```

ch1bo commented on hydra-node/src/Hydra/API/RestServer.hs:
```diff
@@ -20,62 +33,88 @@ deriving stock instance IsTx tx => Show (DraftCommitTxResponse tx)
+  arbitrary = genericArbitrary
+
+newtype DraftCommitTxRequest = DraftCommitTxRequest
+  { utxos :: [DraftUTxO]
```
Continuing my comment from above with the JSON serialization, a haskell type fulfilling the more UTxO-like + witness serialization quite directly could be: `UTxO' (TxOut CtxUTxO, Maybe ScriptInfo)`

ch1bo commented on hydra-node/src/Hydra/Chain.hs:
```diff
@@ -186,12 +194,16 @@ data Chain tx m = Chain
+  -- are handling the following with 400 responses: 'CannotFindOwnInitial',
+  -- 'CannotCommitReferenceScript', 'CommittedTooMuchADAForMainnet',
+  -- 'UnsupportedLegacyOutput' and other possible exceptions are turned into 500
+  -- errors
```
This has nothing to do with HTTP, so do not speak about 400 or 500 status codes here. Instead, document that behavior on the API call site if you want.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Handlers.hs:
```diff
@@ -145,11 +162,27 @@ mkChain tracer queryTimeHandle wallet ctx LocalChainState{getLatest} submitTx =
+                  & addInputs
+                    ( second (BuildTxWith . ScriptWitness ScriptWitnessForSpending)
+                        <$> scriptWitnesses
+                    )
```
This should stay very simple code without much additional logic (maybe except the checking for wallet utxo).

To achieve that, do not build the transaction half-way and extend it here, but pass in a more complete type right away. That is, not only `UTxO` which is "regular utxo" as you call it, but using `UTxO' (TxOut CtxUTxO, Maybe ScriptWitness)` (or something along those lines - a conversion from `ScriptInfo` to the cardano-api ScriptWitness would make sense maybe on the API layer already).

Then, in the `Direct.Tx` module, we change the signature of `commitTx` to take this `UTxO' (TxOut CtxUTxO, Maybe ScriptWitness)` (we could think about a type alias `type UTxOToCommit = UTxO' (TxOut CtxUTxO, Maybe ScriptWitness)`) and convert that accordingly to `TxIns` to construct the right transaction body.

ffakenz replied:
We are not longer building the transaction half-way. Instead we are passing a more complex type as requested and made the mapping `ScriptInfo -> ScriptWitness` also in the API layer [[commit](https://github.com/input-output-hk/hydra/pull/903/commits/e52265e1ed106ddd1daa5537b801f104ca930a54)].

Still missing changes in `Direct.Tx` module tho.

ch1bo commented on hydra-node/test/Hydra/API/RestServerSpec.hs:
```diff
@@ -5,15 +5,36 @@ module Hydra.API.RestServerSpec where
+      withMaxSuccess 1 $ do
+        conjoin
+          [ prop_validateJSONSchema @(DraftCommitTxResponse Tx) "api" (key "channels" . key "/commit" . key "subscribe")
+          , prop_specIsComplete @(DraftCommitTxResponse Tx) "api" apiSpecificationSelector
```
This is a bit pointless as `DraftCommitTxResponse` is a product type and has not multiple variants which could be checked for completeness. However, the `/commit` endpoint should have multiple responses, also in the api.yaml [here](https://github.com/input-output-hk/hydra/pull/903/files#diff-f20909c04bf600ba813ac22890c66583bc03b69c1d210fc2b7b3ff30c8d9fb6eR120-R121)

ch1bo replied:
Should: Specify the possible errors in the `/commit` subscribe endpoint. It should be `oneOf` the nominal response and the possible error responses we send. Looking at https://github.com/input-output-hk/hydra/pull/903/files#diff-f1fbd8fea2154d2538329d3e931106afa899ed70716d79fc58b84ed2a0406b42R387-R390, I would expect at least the 400 responses to be documented in the api.

ffakenz committed:
Remove unnecessary comment on chain module

ffakenz committed:
Replace DraftUTxO with a UTxO like type with witness

Also removed unused golden file.

ffakenz committed:
Add deriving json instance on UTxO'

This avoids the need to define a deriving instance for UTxO at
Hydra.Ledger.Cardano.

This also simplifies the json instances for the api server.

ffakenz committed:
Include an example for a draft commit api request

ffakenz committed:
Fix after rebase

ffakenz committed:
Refactor draftCommitTx interface to take UTxO' like type as argument instead

We do not longer build the transaction half-way and extend it on the handler,
but pass in a more complete type right away.

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -755,19 +738,60 @@ components:
+                  "value": {
+                    "lovelace": 7620669
+                  }
+                },
```
This schema is incomplete. I suspect this example will not even parse as `witness` is not really optional right now.

Also the `$ref: ...` above need to point not to `UTxO`, but to something that is `UTxO` + the `witness` field

ffakenz replied:
Indeed this example is not parseable. 

[Fixed](https://github.com/input-output-hk/hydra/pull/903/commits/b27bf8eb3aa9b10be09cf5be5d855805a34753be) and added new types with witnesses in `api.yaml` to not point to `UTxO`.

Still missing to make `witness` really optional. Hints?

ch1bo replied:
Left a hint on the `FromJSON`, I would suggest you update the example here to remove `"witness":null`

ffakenz committed:
Fix examples for draft commit request in api yaml

Although witness is an optional type, it has to be provided as null.
Also create new types with witness to avoid using the regular UTxO api type:
- UTxOWithWitness
- TxOutWithWitness
- Witness

ffakenz commented on hydra-node/src/Hydra/API/Server.hs:
```diff
@@ -370,9 +387,11 @@ handleDraftCommitUtxo directChain tracer body reqMethod reqPaths respond = do
               CannotCommitReferenceScript -> return400 e
               CommittedTooMuchADAForMainnet _ _ -> return400 e
               UnsupportedLegacyOutput _ -> return400 e
+              walletUtxoErr@SpendingNodeUtxoForbidden -> return400 walletUtxoErr
```
ErrCoverFee are not covered and considered 500 status code, is that a problem?

ch1bo replied:
Unless we run into lookup problems before, maybe the `ErrUnknownInput` could be raised due to a user picking a `TxIn` which is unknown.

So yes, I would say that error case would also be a 400

pgrange reviewed (approved):
I approve and share som feedbacks above.

Appreciate:
* the work done here
* the test that ensure that the signed transaction can be posted in the end

To make it perfect I would improve error handling, in particular:
* split the type`ErrCoverFee` and return 400 for:
  + ErrUnknownInput
  + ErrScriptExecutionFailed
* split the type `ErrCoverFee` and return 500 for:
  + ErrNotEnoughFunds (as the node is supposed to provide for the fees)
  + ErrNoFuelUTxOFound (but maybe this will disappear)
  + ErrTranslationError (but honestly I don't know what this one mean)
* Split the type `PostTxError` end/or rename it since we don't post any tx here and make it so that we use a smaller type for this case
* Split the type `PostTxError` end/or rename it and return 500 for:
  + NoSeedInput
  + CannotFindOwnInitial
  + InternalWalletError
* Split the type `PostTxError` end/or rename it and return 400 for:
  + UnsupportedLegacyOutput
  + ScriptFailedInWallet (we suppose it's the user request fault that the script failed)
  + PlutusValidationFailed (we suppose it's the user request fault that the plutus validation failed)
  + CommittedTooMuchADAForMainnet
  + FailedToDraftTxNotInitializing
* Split the type `PostTxError` end/or rename it and remove:
  + CannotCommitReferenceScript (I guess we can now)
* Split the type `PostTxError` end/or rename it and remove, for this case:
   + InvalidStateToPost (makes no sense if we don't post tx)
   + NotEnoughFuel (covered by ErrNotEnoughFunds above?)
  + NoFuelUTXOFound (covered by ErrNotEnoughFunds above?)
  + FailedToPostTx (we don't post here)

ch1bo commented on hydra-node/src/Hydra/API/RestServer.hs:
```diff
@@ -20,62 +43,105 @@ deriving stock instance IsTx tx => Show (DraftCommitTxResponse tx)
-      [ "tag" .= String "DraftCommitTxRequest"
-      , "utxos" .= toJSON utxo
-      ]
+data DraftUTxO = DraftUTxO
```
Could: call this `DraftTxOut` or as it is actually an enriched `TxOut`

ffakenz replied:
Indeed it is an enriched `TxOut`.

Renaming to `TxOutWithWitness`.

ch1bo commented on hydra-node/src/Hydra/API/RestServer.hs:
```diff
@@ -20,62 +43,105 @@ deriving stock instance IsTx tx => Show (DraftCommitTxResponse tx)
+  parseJSON v = do
+    txOut <- parseJSON v
+    flip (withObject "DraftUTxO") v $ \o -> do
+      witness <- o .: "witness"
```
Should: parse this optionally using [.:?](https://hackage.haskell.org/package/aeson-2.1.2.1/docs/Data-Aeson.html#v:.:-63-)

ch1bo commented on hydra-cardano-api/src/Hydra/Cardano/Api/ScriptData.hs:
```diff
@@ -106,3 +106,8 @@ instance ToJSON HashableScriptData where
 instance FromJSON HashableScriptData where
   parseJSON = fmap unsafeHashableScriptData . parseJSON
+
+-- XXX: Incomplete arbitrary instance
```
Should: not leave this out as it's quite easy to do

Just pattern match on all constructors of `ScriptData` and recursively call `arbitrary` for all arguments

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -171,6 +206,162 @@ singlePartyCommitsUsingFuel tracer workDir node hydraScriptsTxId =
+      let regularUTxO1 = mkDraftUTxOs regularUtxo1 Nothing
+      let regularUTxO2 = mkDraftUTxOs regularUtxo2 Nothing
+      let regularUtxosInfo = regularUTxO1 <> regularUTxO2
+      let clientPayload = DraftCommitTxRequest (regularUtxosInfo <> scriptUtxosInfo)
```
Is there a particular reason why you chose to commit 4 UTxOs (2 regular and 2 scripts)?

If you keep 4, it would be much clearer if you would `<>` all 4 here to highlight it's 4 UTxOs.

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -171,6 +206,162 @@ singlePartyCommitsUsingFuel tracer workDir node hydraScriptsTxId =
+            (ReqBodyJson clientPayload)
+            (Proxy :: Proxy (JsonResponse (DraftCommitTxResponse Tx)))
+            (port $ 4000 + hydraNodeId)
+      responseStatusCode response `shouldBe` 200
```
This line is not needed, `runReq` does throw on exceptions.

Could: Generalize [externalCommit](https://github.com/input-output-hk/hydra/pull/924/files#diff-a659e7df20ade6003820a737c7203bef03512eafba15e6305696ea46e66b6d9dR166) and use it here.

Bonus: Improve `externalCommit` (or the variant you created) to catch/rethrow an exception with a call stack using `HasCallStack` and `failure` to make test errors easier to decipher and know which commit failed.

v0d1ch replied:
For some reason I am getting _redundant constraint_ warning for `HasCallStack`

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -171,6 +206,162 @@ singlePartyCommitsUsingFuel tracer workDir node hydraScriptsTxId =
+  RunningNode ->
+  TxId ->
+  IO ()
+singlePartyCommitsFromExternal tracer workDir node hydraScriptsTxId =
```
Should: remove this test as all other e2e tests are doing this already

In https://github.com/input-output-hk/hydra/pull/924 we have made most tests use external commit of `Normal` outputs, so this tests is redundant.

ffakenz replied:
I know it was redundant but did not know if it was valuable to keep.
Removing.

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -171,6 +206,162 @@ singlePartyCommitsUsingFuel tracer workDir node hydraScriptsTxId =
+  RunningNode ->
+  TxId ->
+  IO ()
+singlePartyCantCommitExternallyWalletUtxo tracer workDir node hydraScriptsTxId =
```
Could: use `Cannot` in a name as it's easier to read and tell apart from `Can` (`Cant` is easily misread)

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -171,6 +206,162 @@ singlePartyCommitsUsingFuel tracer workDir node hydraScriptsTxId =
+            (Proxy :: Proxy (JsonResponse (DraftCommitTxResponse Tx)))
+            (port $ 4000 + hydraNodeId)
+        )
+        `shouldThrow` selector400
```
Not sure if I like `req` after all, this way of matching on errors looks awful. I also did not see a way to `runReq` differently to not have the response body handling with exception throwing.

Using `http-client`/`http-conduit` this could be something like:

```suggestion
      res <- httpNoBody $ "POST http://127.0.0.1/commit"
               & setRequestPort (4000 + hydraNodeId)
               & setRequestBodyJSON clientPayload
      getResponseStatusCode res `shouldBe` 400
```

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -210,6 +401,10 @@ canCloseWithLongContestationPeriod tracer workDir node@RunningNode{networkId} hy
     (fuelUTxO, otherUTxO) <- queryMarkedUTxO node actorVk
     traceWith tracer RemainingFunds{actor = actorName actor, fuelUTxO, otherUTxO}
 
+mkDraftUTxOs :: UTxO -> Maybe ScriptInfo -> UTxO' DraftUTxO
```
This is a weird signature and it's semantics will pair the script info to each entry of the `UTxO`, which could be surprising.

Could: Make this a proper smart constructor with signature `mkDraftUTxO :: TxIn -> TxOut -> Maybe ScriptInfo -> UTxO' DraftUTxO` or so.

If this should be hard to use because the `createOutputAtAddress` returns `UTxO`, change that function to return `(TxIn, TxOut)` and do `mkDraftUTxO :: (TxIn, TxOut, Maybe ScriptInfo) -> UTxO' DraftUTxO` instead.

v0d1ch replied:
We use `mkDraftUTxO` in conjunction with `seedFromFaucet` also and `createOutputAtAddress` is used only twice so maybe not worth the change.

ch1bo commented on hydra-cluster/src/HydraNode.hs:
```diff
@@ -205,6 +206,7 @@ data EndToEndLog
   | RemainingFunds {actor :: String, fuelUTxO :: UTxO, otherUTxO :: UTxO}
   | PublishedHydraScriptsAt {hydraScriptsTxId :: TxId}
   | UsingHydraScriptsAt {hydraScriptsTxId :: TxId}
+  | SubmitCommitTx {commitTx :: Tx}
```
Must: remove unused data constructor

ffakenz replied:
Nice catch!

ch1bo commented on hydra-node/golden/ReasonablySized DraftCommitTxRequest.json:
```diff
@@ -0,0 +1,212 @@
+                        },
+                        "lovelace": 17
+                    },
+                    "witness": null
```
Should: not serialize `Nothing` as `null` entry here.

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -755,19 +738,60 @@ components:
+      title: DraftCommitTxRequest
+      description: |
+        Provide utxos that will be used to draft a commit transaction that will be sent back to the user.
+        Script utxos must provide a witness, regulars can ommit this.
```
```suggestion
        Script utxos must provide a witness, regulars can omit this.
```

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -755,19 +738,60 @@ components:
+        additionalProperties: false
+        properties:
+          utxos:
+            type: array
```
Must: fix the schema, `utxos` is *NOT* an array, but an object where keys are `TxIn`

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -1564,6 +1600,65 @@ components:
+    TxOutWithWitness:
+      type: object
+      description: |
+        A single transaction output
```
```suggestion
        A single transaction output with optional witness to spend it.
```

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -1599,6 +1694,67 @@ components:
+          redeemer:
+            type: string
+          datum:
+            type: string
```
Should: use `$ref: "#/components/schemas/Cbor"` to be more precise what kind of string is used for `redeemer` and `datum`

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -1599,6 +1694,67 @@ components:
+      - type
+      properties:
+        cborHex:
+          type: string
```
```suggestion
          $ref: "#/components/schemas/Cbor"
```

ch1bo commented on hydra-node/src/Hydra/API/RestServer.hs:
```diff
@@ -1,13 +1,36 @@
+import Hydra.Ledger (IsTx)
+import Hydra.Ledger.Cardano ()
 
 newtype DraftCommitTxResponse tx = DraftCommitTxResponse
```
Should: make this also not parameterized by `tx` anymore

As we have seemingly have dropped the `tx` on the `DraftCommitTxRequest`, we should also make this specific to cardano transactions by using `commitTx :: Tx`. As a consquence the golden / roundtrip tests need to be updated and a golden file needs to be deleted.

ch1bo replied:
This would even address #945 if we define

```hs
toJSON (DraftCommitTxResponse tx) =
    object
      [ "commitTx" .= (String . decodeUtf8 . Base16.encode $ serialiseToCBOR tx)
      ]
```

and `FromJSON` using `deserialiseFromCBOR`.

But then, the other `ToJSON` occurrences where instances of `Tx` are used are still broken/different. CC @v0d1ch

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/State.hs:
```diff
@@ -313,57 +318,75 @@ commit ctx st utxo = do
+    Just initial -> do
+      rejectByronAddress utxo
+      rejectReferenceScripts utxo
+      rejectMoreThanMainnetLimit networkId utxo
```
Must: not duplicate logic when it is not needed, but update the signature of `commit` above instead.

```
commit ::
  ChainContext ->
  InitialState ->
  UTxO' (TxOut CtxUTxO, Maybe (ScriptWitness WitCtxTxIn Era)) ->
  Either (PostTxError Tx) Tx
```

could work and should simplify the call site as well.

ch1bo commented on hydra-node/src/Hydra/API/RestServer.hs:
```diff
@@ -20,62 +43,105 @@ deriving stock instance IsTx tx => Show (DraftCommitTxResponse tx)
+  toScriptWitness ScriptInfo{redeemer, datum, plutusV2Script} =
+    mkScriptWitness plutusV2Script (ScriptDatumForTxIn datum) redeemer
+
+prepareCommitTxInputs ::
```
If we change interfaces as I have outlined in other comments, this should not be needed.

ch1bo commented on hydra-node/test/Hydra/JSONSchema.hs:
```diff
@@ -23,31 +23,79 @@ import System.Exit (ExitCode (..))
 -- the specification. Note this, because this uses an external tool each
 -- property iteration is pretty slow. So instead, we run the property only
 -- once, but on a list of 100 elements all arbitrarily generated.
+-- TODO: This should be replacable by 'prop_validateJSONSchema'
```
Should: resolve TODO (or explain why not as per PR checklist)

ch1bo replied:
Done in https://github.com/input-output-hk/hydra/commit/a9324ebadc482520d6acc3cbc8eb9d11dca70a02

ch1bo commented on hydra-node/test/Hydra/JSONSchema.hs:
```diff
@@ -132,22 +182,26 @@ type SpecificationSelector = Traversal' Aeson.Value Aeson.Value
+        let spec' = addField "$id" ("file://" <> dir <> "/") spec
+        liftIO $ Aeson.encodeFile (dir </> takeBaseName file <.> "json") spec'
+        -- XXX: We need to write the specFile as .yaml as well because internally the
+        -- spec .json reference elements within the yaml file.
```
```suggestion
        -- XXX: We need to write the specFile as .yaml although it is a JSON document now, 
        -- because internally the spec reference elements using the original .yaml file name.
```

ch1bo commented on hydra-node/test/Hydra/LoggingSpec.hs:
```diff
@@ -24,14 +23,13 @@ spec = do
+      withMaxSuccess 1 $
+        conjoin
+          [ prop_validateToJSON @(Envelope (HydraLog Tx ())) "logs" "messages" "HydraLog"
+          , prop_specIsComplete @(HydraLog Tx ()) "logs" apiSpecificationSelector
```
Should: not request half file names. `"logs"` will be expanded to `logs.yaml` inside this function and this is weird. If we would keep `"logs.yaml"` as before then it is clear(er) that the file in our working tree is meant.

ch1bo commented on hydra-node/src/Hydra/Chain.hs:
```diff
@@ -143,6 +149,8 @@ data PostTxError tx
     CommittedTooMuchADAForMainnet {userCommittedLovelace :: Lovelace, mainnetLimitLovelace :: Lovelace}
   | -- | We can only draft commit tx for the user when in Initializing state
     FailedToDraftTxNotInitializing
+  | -- | User should not be able to draft a commit tx using the wallet utxo
```
```suggestion
  | -- | Committing UTxO addressed to the internal wallet is forbidden.
```

ch1bo commented on hydra-cluster/hydra-cluster.cabal:
```diff
@@ -133,6 +133,8 @@ library
     , filepath
     , hspec-expectations
+    , http-conduit
+    , http-types
```
Are we really using both to match the `HttpException`?

v0d1ch replied:
I removed both and used http-client instead since it contains all we need.

ch1bo reviewed (changes requested):
Many small possible improvements, which are not always required to be done right now.

The schema still seems to be incomplete, and the internal code duplication is something that must be addressed in my opinion (marked "Must" changes)

ffakenz committed:
Update hydra-node/json-schemas/api.yaml

Co-authored-by: Sebastian Nagel <ch1bo@users.noreply.github.com>

ffakenz committed:
Update hydra-node/src/Hydra/Chain.hs

Co-authored-by: Sebastian Nagel <ch1bo@users.noreply.github.com>

ffakenz committed:
Update hydra-node/test/Hydra/JSONSchema.hs

Co-authored-by: Sebastian Nagel <ch1bo@users.noreply.github.com>

ffakenz committed:
Update hydra-node/json-schemas/api.yaml

Co-authored-by: Sebastian Nagel <ch1bo@users.noreply.github.com>

ffakenz committed:
Update hydra-node/json-schemas/api.yaml

Co-authored-by: Sebastian Nagel <ch1bo@users.noreply.github.com>

ffakenz committed:
Apply first round of request for review

- Rename DraftUTxO to TxOutWithWitness.
- Fix witness parsing for TxOutWithWitness to be really optional.
This also fixes the golden file for DraftCommitTxRequest.
- Define arbitrary instance for ScriptData.
- Complete arbitrary instance for HashableScriptData.
- Remove singlePartyCommitsFromExternal scenario as it was redundant.
In #924 we have made most tests use external commit of Normal outputs.
- use Cannot in a scenario name as it's easier to read and tell apart from Can.
Cant is easily misread.
- Remove unused data constructor SubmitCommitTx.
- use $ref: "#/components/schemas/Cbor" for Script datum and redemer in api.yaml

v0d1ch committed:
Remove http-conduit and http-types in favor of http-client

ffakenz committed:
Pass full file name with extension to JSONSchema property validations

v0d1ch committed:
Make externalCommit function accept `UTxO' TxOutWithWitness`

- Catch `HttpException` in the externalCommit and rethrow

v0d1ch committed:
Be clear we are using 4 utxos

v0d1ch committed:
WIP: remove rawCommitTxBody function

ffakenz committed:
Generate script inputs arbitrarily

v0d1ch commented:
@pgrange Related to your comment [here](https://github.com/input-output-hk/hydra/pull/903#pullrequestreview-1493497875) - we talked about this separation at some point with @ch1bo and we agreed that we would need to re-model error handling since as you saw there are some redundancies here and there and _translating_ from one type to another so let's leave this out of scope for now if you agree?

v0d1ch committed:
Alter the api.yaml type for utxos

v0d1ch committed:
Remove abstract tx type param and use a fixed Tx type in DraftCommitTxResponse

v0d1ch committed:
Document possible responses

Also remove tests that were not covering all cases properly

ch1bo commented on hydra-cardano-api/src/Hydra/Cardano/Api/ScriptData.hs:
```diff
@@ -96,13 +98,35 @@ instance FromJSON ScriptData where
+   where
+    -- XXX: ScriptDataBytes expects utf-8 words
+    arbitraryBS =
+      BS.pack <$> vectorOf 8 arbitrary
```
Script data will always be shorter than 64 bytes, but can also be empty or anything between.

```suggestion
    arbitraryBS = BS.pack <$> choose (0, 64) >>= vector
```

v0d1ch committed:
PR review comments

ch1bo commented on hydra-node/bench/tx-cost/TxCost.hs:
```diff
@@ -102,7 +102,9 @@ computeCommitCost = do
     (cctx, stInitial) <- genStInitial ctx
-    pure (commit cctx stInitial utxo, getKnownUTxO stInitial <> getKnownUTxO cctx)
+    -- TODO: should we generate script inputs to calculate the costs more
+    -- realistically or probably have a separate function to do that?
```
No, remove the TODO

ch1bo commented on hydra-cluster/src/HydraNode.hs:
```diff
@@ -163,19 +165,27 @@ waitForAll tracer delay nodes expected = do
 -- | Create a commit tx using the hydra-node for later submission
-externalCommit :: HydraClient -> UTxO -> IO Tx
+-- Only works for non-script utxos
+externalCommit :: HasCallStack => HydraClient -> UTxO' TxOutWithWitness -> IO Tx
```
Not complicate the generic case of normal UTxO committing, but instead define a `externalCommit'` which takes that more complicated type.

ch1bo committed:
Change signature of commitTx and commit functions

These expand the original UTxO argument to a UTxO including witnesses.
This way, we can pack the script UTxOs into the same value and the
commit transaction creation will just work™.

ch1bo committed:
Simplify commit scripts from external e2e scenario

There is no point in mixing multiple utxo into such an integrated
scenario. Let's keep the test case on point for committing a script into
a head end-to-end. Should we desire more combinations, a "lower level"
test suite should be exercised.

ch1bo committed:
Fix ScriptData instances

Note that it's note safe to use unsafeHashableScriptData when the data
was already serialized (as in the FromJSON instance).

ch1bo commented on hydra-node/src/Hydra/API/Server.hs:
```diff
@@ -337,8 +358,7 @@ instance Exception RunServerException
 handleDraftCommitUtxo ::
-  forall tx.
-  IsChainState tx =>
+  -- IsChainState tx =>
```
Should: remove Commented code

ch1bo committed:
Use prop_validateJSONSchema in HydraLoggingSpec

Seemingly the previously used prop_validateToJSON was not checking the
spec correctly (anymore) as there were at least two issues in the
logs.yaml specification.

ch1bo reviewed (approved):
This is nearly done! Only the `externalCommit` ergonomy in the tests should not degrade. I do approve despite this already (under the assumption that we do not change the code much anymore)

ffakenz committed:
Remove commented code

v0d1ch committed:
Introduce a wrapper around externalCommit

v0d1ch committed:
Remove HasCallstack

- It caused one test to fail and from the error outcome I could not
see the actuall stack trace printed.
